src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+                                    "image_frame parameter."};
+    }
+    switch (params_.img_frame) {
+    case image_frame::file:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::pipe:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
+                              static_cast<std::size_t>(params_.data_shape[2]),
+                              static_cast<std::size_t>(params_.data_shape[0])}}
+            .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(tsr);
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    auto instance_offset =
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (const instance &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::pipe) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
+        }
+        else {
+            img_buf = ins.bits();
+        }
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (params_.data_shape[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
+                            params_.data_shape[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (params_.resize != -1) {
+            resize(tmp, tmp);
```

Comment:
Is it safe to use the same memory region as both source and destination of the resize operation?

Reply:
So usage looks safe as internally it is saving `src` and re-allocating `dst` based on the code here: https://github.com/opencv/opencv/blob/master/modules/imgproc/src/resize.cpp#L3922

This is how they seem to be using in the referenced examples as well [search for `resize`]:
- https://docs.opencv.org/4.2.0/d0/df8/samples_2cpp_2train_HOG_8cpp-example.html#a55
- https://docs.opencv.org/4.2.0/d4/db9/samples_2dnn_2object_detection_8cpp-example.html#a37 

However, if you think that relying on the method's internal implementation for memory management is bad practice then please do let me know - i'll refactor.